joomla/joomla-cms
# Language associations improvements

#### Description

This PR takes similar languages associations code from language filter, mod languages and some helpers and centralize it in `JLanguageAssociations`.
It also ...
###### Does some improvements

The components without menu item id now will have an alternate language association too.
For instance, the if current URI is a component without menu item (ex: /en/component/users/), the associated language URI will be the version of the component page in the other language (ex: /fr/component/users/).
###### Corrects some bugs
1. If you have 2 languages (published and public) but one was the home menu item id with a higher view level. It shouldn't appear in the language meta tags nor in mod languages flags/text.
2. If you have 2 languages (published and public) and the 2 home item menus (also published and public) but the language ini folder/file does not exist. It shouldn't appear in the language filter nor in mod languages flags/text.
###### Performance

Also there is a very slight performance improvement in a multilanguage site, because the language switcher and the language filter plugin now use the same method to get the languages associations, instead of doing that two times in each page.
##### New method JLanguageMultilang::getAvailableLanguages()

Besides the new methods in `JLanguageAssociations`, this introduces a new method `JLanguageMultilang::getAvailableLanguages()` that is different from the `JLanguageHelper::getLanguages()`.

`JLanguageHelper::getLanguages()` only gets the content languages that are published.

`JLanguageMultilang::getAvailableLanguages()` uses `JLanguageHelper::getLanguages()` to get the content languages published and them does some checks to check if that language is really available to the user, it checks:
- if the language extension exists and is enabled. Though `JLanguageMultilang::getSiteLangs()`
- if the user has view level rights to that published language
- if the language has a published homepage. Though `JLanguageMultilang::getSiteHomePages()`
- if the user has view level rights to that published homepage
- if there is a ini file for that language

And also returns the language object with the language menu item homepage id.

See https://github.com/andrepereiradasilva/joomla-cms/blob/language-association-improvement/libraries/cms/language/multilang.php#L136-L176
#### How to test
1. Install latest staging with multilanguage. Language filter on, 2 or more languages with language homepages and language switcher module on.
2. Install this patch.
3. Make some language associations.
4. Test language associations with or without language published, different view levels, etc. Also, different language filter plugin options.
   Check if the language switcher and the alternate meta code in the head are as they should.
5. Check if bugs described above are corrected.
6. Check if component without menu items now have the language association (ex: /en/component/users/).
#### Observations

Improvements, suggestions and code revisions are welcome.

@infograf768, please check this one, when you have time.
